### PR TITLE
feat(quantic): make quantic notifications component dismissible 

### PR DESCRIPTION
**Pull request by SimonMilord**
[SFINT-5767](https://coveord.atlassian.net/browse/SFINT-5767)

## IN THIS PR:
- Fixed a styling issue we had with the notifications where the width would look weird.
- Made the quantic notifications component dismissible, thus following best practices by Salesforce and offering a better user experience.
- Made sure: that if you dismiss notifications and re-trigger the same query, that they would re-appear
- Made sure: that if you dismiss notifications and trigger a new search, the previous notifications would not "flash" before updating the state

More about the solution:

The `QuanticNotifications` component was rendering the notifications based on the notifications array it was getting from the notify trigger state from headless. We are therefore adding a button to close the notification in the template, which will call the `handleNotificationClose` handler function. 

## DEMO STYLING:

Before:
Screenshot 2024-11-27 at 4 59 35 PM

After:
image

## DEMO FUNCTIONALITY:

https://github.com/user-attachments/assets/c9aa3ed5-8468-4192-88bb-2ee0ea254449

## BONUS:

Resized the default icon for x close button to 16px instead of 24px which looks nicer:

image

## TESTS:

image

[SFINT-5767]: https://coveord.atlassian.net/browse/SFINT-5767?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by mmitiche** [comment]
Visual review: 

Nice work 👍 
I think the dismiss button a bit huge, it will look much better in my opinion if we make it smaller, like the size of the button to clear the query in the search box for example.

**Comment by SimonMilord** [comment]
> Visual review:
> 
> Nice work 👍 I think the dismiss button a bit huge, it will look much better in my opinion if we make it smaller, like the size of the button to clear the query in the search box for example.

Set the height of the icon to 16px which looks nicer! thanks

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js [hunk 1]
```diff
@@ -56,8 +56,10 @@ export default class QuanticNotifications extends LightningElement {
     this.unsubscribe?.();
   }
 
-  updateState() {
-    this.notifyTriggerState = this.notifyTrigger.state;
+  updateState(updatedNotifications) {
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js:59]: I don't really like what's being done with this function,  as in I don't like that this function can now support having a parameter or not, 

before this change this function had one clear job, updating the component state when the Headless state changes, clear and concise now it's not only that, and I think we can achieve the same result in a simpler way.


Plus I don't think it's a good idea to kinda separate the Headless state from the component state, with this logic added in this PRs it will be possible to have for example in the Headless state: `['notification 1', 'notification 2']` and in the component state just ['notification 1'].

What would be much better I think than removing data is to augment data with a key that indicates whether the notification is visible or not. for example the state in Headless would be `['notification 1', 'notification 2']`, and the state in the component would be:`[{value: 'notification 1', visible: true}, {value:'notification 2', visible: false}]`

I think this way of doing would yield a simpler code, easy to read and easier to understand,
It would be something like this:

```javascript
  updateState() {
    this.notifications = this.notifyTrigger?.state?.notifications.map((notification, index) => ({
      value: notification,
      id: index.toString(),
      visible: true,
    })) ?? [];
  
    this.ariaLiveNotificationsRegion.dispatchMessage(
      this.notifyTriggerState?.notifications.reduce(
        (value, notification, index) => {
          return `${value} Notification ${index + 1}: ${notification}`;
        },
        ''
      )
    );
  }

  handleNotificationClose(event) {
    const currentNotificationId = event.currentTarget.dataset.id;
    this.notifications = this.notifications.map((notification) => {
      if (notification.id === currentNotificationId) {
        return { ...notification, visible: false }
      }
      return notification
    });
  }

then in the HTML you only display a notification only when visible is true.
This way I think at first glance you can understand everything going on with this component.
```

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.html [hunk 2]
```diff
@@ -12,6 +12,20 @@
           <div class="slds-notify__content slds-var-m-left_small">
             {notification.value}
           </div>
+          <div class="slds-notify__close">
+            <button
+              class="slds-button slds-button_icon slds-button_icon-inverse"
+              title="Close"
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.html:18]: This button needs a better title so when the screen reader reads it, the user understands really what this button is about.

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js (hunk 1 above)
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js:59]: yeah good point!

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js [hunk 3]
```diff
@@ -68,13 +81,25 @@ export default class QuanticNotifications extends LightningElement {
...
-    );
+  handleSearchStatusChange() {
+    if (!this.searchStatus.state.isLoading) {
+      this.notifications =
+        this.notifyTrigger?.state?.notifications.map((notification, index) => ({
+          value: notification,
+          id: index.toString(),
+          visible: true,
+        })) ?? [];
+    }
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js:92]: I would do that, that way on each "new query" the notifications are actually cleared so for example they don't display while the rest of the UI is in "loading" mode, aka showing placeholders.

```suggestion
    if (this.searchStatus?.state?.isLoading) {
      this.notifications = [];
    } else {
      this.notifications =
        this.notifyTrigger?.state?.notifications.map((notification, index) => ({
          value: notification,
          id: index.toString(),
          visible: true,
        })) ?? [];
    }
```

Also why did you add the `toString()` ?

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.html [hunk 4]
```diff
@@ -5,15 +5,31 @@
...
+                title="Close notification"
+                onclick={handleNotificationClose}
+                data-id={notification.id}
+              >
+                <lightning-icon
+                  icon-name="utility:close"
+                  alternative-text="Close"
+                  size="x-small"
+                ></lightning-icon>
+              </button>
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.html:28]: Is there a reason why we're not using the `lightning-button-icon icon-name="utility:close" variant="inverse"` etc. ?

https://developer.salesforce.com/docs/component-library/bundle/lightning-button-icon/example

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.css [hunk 5]
```diff
@@ -1,7 +1,15 @@
+:host {
+  --slds-c-toast-sizing-min-width: 100%;
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.css:2]: Why is this necessary?

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.css [hunk 6]
```diff
@@ -1,7 +1,15 @@
...
 .notification {
   background-color: var(---lwc-colorBackgroundAlt, white);
   border: 1px solid var(--lwc-colorBackgroundToast, rgb(116, 116, 116));
   margin-right: 0;
   margin-left: 0;
   width: 100%;
+}
+
+.notification-container {
+  z-index: 1;
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.css:14]: I personally would put this at like 10?
So at least there can be content in between 0 and the notifications?

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/__tests__/quanticNotifications.test.js [hunk 7]
```diff
@@ -97,6 +113,12 @@ function mockErroneousHeadlessInitialization() {
   };
 }
 
+// Simulates a search status update
+function mockSearchStatusUpdate() {
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticNotifications/__tests__/quanticNotifications.test.js:117]: Either you put this directly in your test, or you find it a better name, in the context we have when reading that function name, we have no idea what you're updating the searchStatus to.

`mockSearchStatusHasResults` for example describes what you're doing now.

Otherwise I would simply put the code within your tests specifically since what you update the state to can be very specific to the test you're running.

**Review by erocheleau — COMMENTED** [review]
Some small improvements that I think can help clarity

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.css (hunk 5 above)
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.css:2]: I documented that in the PR description, it is to fix the width issue of the notifications

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.css (hunk 6 above)
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.css:14]: modal is 40 so we could do steps of 20 and set it at 20, but I didnt think there would be content between 0 and 1

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/__tests__/quanticNotifications.test.js (hunk 7 above)
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticNotifications/__tests__/quanticNotifications.test.js:117]: will change naming

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.html (hunk 4 above)
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.html:28]: no particular reasons, I could change it also

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js (hunk 3 above)
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js:92]: I will test the first point you bring up, seems to make sense

for the second question, it is because in the `handleNotificationClose` when we do: `if (notification.id === currentNotificationId)`, the currentNotificationId is of type string, so it was so that they could do the comparison, but when I think about it, I could probably just do `if (notification.id.toString() === currentNotificationId)`

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.css [hunk 8]
```diff
@@ -1,7 +1,15 @@
...
 .notification {
   background-color: var(---lwc-colorBackgroundAlt, white);
   border: 1px solid var(--lwc-colorBackgroundToast, rgb(116, 116, 116));
   margin-right: 0;
   margin-left: 0;
   width: 100%;
+}
+
+.notification-container {
+  z-index: 20;
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.css:14]: z-index is not good with the expandable search box:

https://github.com/user-attachments/assets/98464d13-2fc8-480e-bae2-3af6246a4401

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.html [hunk 9]
```diff
@@ -5,15 +5,28 @@
...
+            <div class="slds-notify__content slds-var-m-left_small">
+              {notification.value}
+            </div>
+            <div class="slds-notify__close">
+              <lightning-button-icon
+                icon-name="utility:close"
+                alternative-text="Close"
+                size="small"
+                variant="bare"
+                title="Close notification"
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.html:22]: This label needs to be localized.

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js [hunk 10]
```diff
@@ -33,6 +34,12 @@ export default class QuanticNotifications extends LightningElement {
   hasInitializationError = false;
   /** @type {import('c/quanticUtils').AriaLiveUtils} */
   ariaLiveNotificationsRegion;
+  /** @type {Array} */
+  notifications = [];
+  /** @type {Function} */
+  unsubscribe;
+  /** @type {Function} */
+  unsubscribeSearchStatus;
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js:42]:
```suggestion
  unsubscribeSearchStatus;
  /** @type {SearchStatus} */
  searchStatus;
```

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js [hunk 11]
```diff
@@ -68,13 +81,27 @@ export default class QuanticNotifications extends LightningElement {
...
 
-  get notifications() {
-    return (
-      this.notifyTriggerState?.notifications.map((notification, index) => ({
-        value: notification,
-        id: index,
-      })) ?? []
-    );
+  handleSearchStatusChange() {
+    if (this.searchStatus?.state?.isLoading) {
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js:85]: You need to handle the error state as well, as in if an error occurs with a search query, the current notifications should be hidden as well.
this is also something you use from the search status state.
It is something to cover with unit tests as well.
Screenshot 2024-12-03 at 11 06 13 AM

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/__tests__/quanticNotifications.test.js [hunk 12]
```diff
@@ -97,6 +113,12 @@ function mockErroneousHeadlessInitialization() {
   };
 }
 
+// Simulates a search status update
+function mockSearchStatusHasResults() {
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticNotifications/__tests__/quanticNotifications.test.js:117]: You could simply name the function `simulateSearchStatusUpdate` and delete the comment.

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/__tests__/quanticNotifications.test.js [hunk 13]
```diff
@@ -224,4 +247,86 @@ describe('c-quantic-notifications', () => {
       });
     });
   });
+
+  describe('when clicking on a notification close button', () => {
+    it('should properly dismiss that notification', async () => {
+      const element = createTestComponent();
+      mockSearchStatusHasResults();
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticNotifications/__tests__/quanticNotifications.test.js:254]: Why do you need this here?

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.css (hunk 8 above)
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.css:14]: nice catch, the notifications I had tested were under the tabs, so I did not catch this one. 

I changed the z-index from 20 to 2 since the quanticSearchbox is at 5.

This allows the expandable searchbox to be ok and the notifications to still take priority over regular page content which has z-index of 1

image

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/__tests__/quanticNotifications.test.js (hunk 13 above)
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticNotifications/__tests__/quanticNotifications.test.js:254]: because `this.notifications` is set in `handleSearchStatusChange` which is called when the searchStatus changes:

`this.unsubscribeSearchStatus = this.searchStatus.subscribe(() => this.handleSearchStatusChange());` 

So if I want to trigger a "fake search" which sets the notification, I need to mock a searchStatus change

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js (hunk 11 above)
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js:85]: fixed and tested that as well! 👌 


https://github.com/user-attachments/assets/4d9d8711-963f-40b7-8822-540b7c282c91

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/__tests__/quanticNotifications.test.js (hunk 13 above)
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticNotifications/__tests__/quanticNotifications.test.js]: I know but why do you wanna trigger a fake search when the test is about closing a notification?

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js [hunk 14]
```diff
@@ -68,13 +88,30 @@ export default class QuanticNotifications extends LightningElement {
...
+    ) {
+      this.notifications = [];
+    } else {
+      this.notifications =
+        this.notifyTrigger?.state?.notifications.map((notification, index) => ({
+          value: notification,
+          id: index,
+          visible: true,
+        })) ?? [];
+    }
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js:104]: why notifications are set in this subscriber function and not in the notifications subscriber?

For me that's the logic that I see for example in plain english:
- Subscribe to the notifications state and update `this.notifications` each time the state changes.
- Subscribe to search status state to reset the visibility to true when a new search is detected and empty the notifications when an error occurs.

this would be translated to this:
```javascript
  updateState() {
    this.notifications =
      this.notifyTrigger?.state?.notifications.map((notification, index) => ({
        value: notification,
        id: index,
        visible: true,
      })) ?? [];
  }

  handleSearchStatusChange() {
    if (!this.searchStatus?.state.isLoading) {
      if (this.searchStatus?.state?.hasError) {
        this.notifications = [];
      } else {
        this.notifications =
          this.notifications?.map((notification) => ({
            ...notification,
            visible: true,
          })) ?? [];
      }
    }
  }
```

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js [hunk 15]
```diff
@@ -48,16 +62,22 @@ export default class QuanticNotifications extends LightningElement {
...
   disconnectedCallback() {
     this.unsubscribe?.();
+    this.unsubscribeSearchStatus?.();
   }
 
   updateState() {
-    this.notifyTriggerState = this.notifyTrigger.state;
+    this.notifyTriggerState = this.notifyTrigger?.state;
+
     this.ariaLiveNotificationsRegion.dispatchMessage(
```
[the remark this answers is not in the document]
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js:81]: adressed that!

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/__tests__/quanticNotifications.test.js (hunk 13 above)
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticNotifications/__tests__/quanticNotifications.test.js]: following discussions, we agreed on a solution, the search status is needed for 2 cases: when you dismiss notif and re-do the same query, it should make the same notifs re-appear and also because of a weird glitch when dismissing notifs and triggering another query where the 'old notifs' would appear for a second before the state was updated.

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js [hunk 16]
```diff
@@ -48,16 +62,58 @@ export default class QuanticNotifications extends LightningElement {
...
+      if (notification.id.toString() === currentNotificationId) {
+        return {...notification, visible: false};
+      }
+      return notification;
+    });
+  }
+
+  dispatchAriaLiveNotificationsRegionMessage() {
     this.ariaLiveNotificationsRegion.dispatchMessage(
       this.notifyTriggerState?.notifications.reduce(
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js:118]: This could use `this.notifications` directly instead of `this.notifyTriggerState`.

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js [hunk 17]
```diff
@@ -48,32 +59,60 @@ export default class QuanticNotifications extends LightningElement {
...
+      })) ?? [];
+
+    this.dispatchAriaLiveNotificationsRegionMessage();
+  }
+
+  handleSearchStatusChange() {
+    if (!this.searchStatus?.state.isLoading) {
+      if (this.searchStatus?.state?.hasError) {
+        this.notifications = [];
+      } else {
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js:90]: Is there really a case where you can have notifications but the search state is in error?

Or is that to cover the "in-between" where you could have an error but still have the notifications from the previous search?

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js [hunk 18]
```diff
@@ -48,32 +59,60 @@ export default class QuanticNotifications extends LightningElement {
...
+          })) ?? [];
+        this.dispatchAriaLiveNotificationsRegionMessage();
+      }
+    }
+  }
+
+  handleNotificationClose(event) {
+    const currentNotificationId = event.currentTarget.dataset.id;
+    this.notifications = this.notifications.map((notification) => {
+      if (notification.id.toString() === currentNotificationId) {
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js:104]: Why the `.toString()` ? What type is a notification Id?

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js [hunk 19]
```diff
@@ -48,32 +59,60 @@ export default class QuanticNotifications extends LightningElement {
...
+            visible: true,
+          })) ?? [];
+        this.dispatchAriaLiveNotificationsRegionMessage();
+      }
+    }
+  }
+
+  handleNotificationClose(event) {
+    const currentNotificationId = event.currentTarget.dataset.id;
+    this.notifications = this.notifications.map((notification) => {
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js:103]: Very interesting point here, this method that you used will override the entire `this.notifications` state.

Which is good sometimes, bad some other times. And I can't figure out if it's good or bad here :D 

On one side, it's possible that overriding the entire `this.notifications` will "re-draw" all the notifications HTML, so it could be simplified by doing something like this instead 

`this.notifications.find(notification => notification.id === currentNotification).visible = false;`

Although I do believe that it's possible that LWC will **not** refresh the variable reactivity since you'd be changing a value deep within the state that's being watched by LWC to re-draw.

Overall I think it's good to leave as-is, but consider these two options in the future to make sure to make a conscious decision.

In short, do I change the data "in-place" (a mutation) or do I override the complete object with the new state, (immutability). In general Immutability is regarded as a good thing, especially in functional programming. Because it ensures there's no side-effect since you always change the whole object, so something cannot be "watching" the object and not get notified of changes because someone changed a deep value instead of the whole object.

**Review by erocheleau — APPROVED** [review]
Some questions and a very interesting comment :) But overall all good, just take time to answer them before merging

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js (hunk 18 above)
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js:104]: number and currentNotificationId string

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js (hunk 17 above)
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js:90]: yeah, if you make a new search which triggers a search error, the "old" notifications will still appear

File: packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js (hunk 19 above)
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticNotifications/quanticNotifications.js:103]: good point, I will keep this in mind!
